candle-pyo3/candle/candle.pyi

```diff
@@ -0,0 +1,111 @@
+from typing import Any, List, Tuple, Union
```

Comment:
Did you have to type all this manually? I'm a bit worried that this will be hard to maintain when making changes to the bindings, I imagine that there is no way to generate these automatically instead at the moment? (I would have thought that it's not that hard actually as it's easy to introspect module elements and documentations in python)

Reply:
Yeah i created those manually and you are absolutely right, this will be a nightmare to maintain. I looked into generating them automatically (or at least from the `text_signature`), but PyO3 doesn't seam to support this currently. See:

-  https://github.com/PyO3/pyo3/issues/2454
- https://github.com/PyO3/pyo3/issues/1112

This seams to be blocked by CPython directly: https://github.com/python/cpython/pull/101872 

I'm think we should close this PR as there will probably be a lot of changes in candles implementation. This is kind of unfortunate as using the python bindings without type-hints can be a bit tedious. Do you have any suggestions?

Reply:
Up to you, as you're planning on making changes to the pyo3 bits as long as you keep this file in mind, I'm happy with merging it - especially if you think it makes you more productive when trying out the python bindings.

Reply:
Yeah i don't think creating them manually is sustainable in the long run. I probably look into how [tokenizers](https://github.com/huggingface/tokenizers/tree/main) solves this with their [stub.py](https://github.com/huggingface/tokenizers/blob/main/bindings/python/stub.py) file.